node/block_exchange.go

```diff
@@ -0,0 +1,226 @@
+		return err
+	}
+	bExService.syncerStatus.m.Lock()
+	defer bExService.syncerStatus.m.Unlock()
+	bExService.syncerStatus.started = true
+	return nil
+}
+
+func (bExService *BlockExchangeService) tryInitBlockStoreAndStartSyncer(ctx context.Context, trustedBlock *types.Block) {
+	if trustedBlock != nil {
+		if err := bExService.initBlockStoreAndStartSyncer(ctx, trustedBlock); err != nil {
+			bExService.logger.Error("failed to initialize the headerstore and start syncer", "error", err)
+		}
+	}
+}
+
+func (bExService *BlockExchangeService) writeToBlockStoreAndBroadcast(ctx context.Context, signedHeader *types.SignedHeader) {
+	// For genesis header initialize the store and start the syncer
+	if signedHeader.Height() == bExService.genesis.InitialHeight {
+		if err := bExService.blockStore.Init(ctx, signedHeader); err != nil {
+			bExService.logger.Error("failed to initialize header store", "error", err)
+		}
+
+		if err := bExService.syncer.Start(bExService.ctx); err != nil {
+			bExService.logger.Error("failed to start syncer after initializing header store", "error", err)
+		}
+	}
+
+	// Broadcast for subscribers
+	if err := bExService.sub.Broadcast(ctx, signedHeader); err != nil {
+		bExService.logger.Error("failed to broadcast block header", "error", err)
+	}
+}
+
+// OnStart is a part of Service interface.
+func (bExService *BlockExchangeService) Start() error {
+	var err error
+	// have to do the initializations here to utilize the p2p node which is created on start
+	ps := bExService.p2p.PubSub()
+	bExService.sub = goheaderp2p.NewSubscriber[*types.SignedHeader](ps, pubsub.DefaultMsgIdFn, bExService.genesis.ChainID)
+	if err = bExService.sub.Start(bExService.ctx); err != nil {
+		return fmt.Errorf("error while starting subscriber: %w", err)
+	}
+	if _, err := bExService.sub.Subscribe(); err != nil {
+		return fmt.Errorf("error while subscribing: %w", err)
+	}
+
+	if err = bExService.blockStore.Start(bExService.ctx); err != nil {
+		return fmt.Errorf("error while starting header store: %w", err)
+	}
+
+	_, _, network := bExService.p2p.Info()
+	if bExService.p2pServer, err = newP2PServer(bExService.p2p.Host(), bExService.blockStore, network); err != nil {
+		return err
+	}
+	if err = bExService.p2pServer.Start(bExService.ctx); err != nil {
+		return fmt.Errorf("error while starting p2p server: %w", err)
+	}
+
+	peerIDs := bExService.p2p.PeerIDs()
+	if bExService.ex, err = newP2PExchange(bExService.p2p.Host(), peerIDs, network, bExService.genesis.ChainID, bExService.p2p.ConnectionGater()); err != nil {
+		return err
+	}
+	if err = bExService.ex.Start(bExService.ctx); err != nil {
+		return fmt.Errorf("error while starting exchange: %w", err)
+	}
+
+	if bExService.syncer, err = newSyncer(bExService.ex, bExService.blockStore, bExService.sub, goheadersync.WithBlockTime(bExService.conf.BlockTime)); err != nil {
+		return err
+	}
+
+	// Check if the headerstore is not initialized and try initializing
+	if bExService.blockStore.Height() > 0 {
```

Comment:
[non blocking] 
In general I think it is good practice to make use of `isInitialized` type methods here, so that how we are checking the state of initialization is controlled and can be unit tested. Additionally it reduces the surface area for bugs if different methods are checking a state differently.

Reply:
added an `isInitialized `method